### DISCUSSION TITLE
Making the built crate available in the devshell

### DISCUSSION THREAD

**Discussion by Nequo** [ipetkov/crane#96] [Q&A] [answered]
Hi, I am new to nix so apologies if this seems like a basic question. I am trying to setup a flake to use with rustlings and have gotten this far:

```nix

{
  description = "Build rustlings with crane";
  nixConfig.bash-prompt-prefix = "[rustlings] ";
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";

    crane = {
      url = "github:ipetkov/crane";
      inputs.nixpkgs.follows = "nixpkgs";
    };

    flake-utils.url = "github:numtide/flake-utils";
  };

  outputs = { self, nixpkgs, crane, flake-utils, ... }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = import nixpkgs {
          inherit system;
        };

        rustlings = crane.lib.${system}.buildPackage {
          src = ./.;
        };
      in
      {
        checks = {
          inherit rustlings;
        };

        packages.default = rustlings;

        apps.default = flake-utils.lib.mkApp {
          drv = rustlings;
        };

        devShells.default = pkgs.mkShell {
          inputsFrom = builtins.attrValues self.checks;

          # https://discourse.nixos.org/t/rust-src-not-found-and-other-misadventures-of-developing-rust-on-nixos/11570/5
          # RUST_SRC_PATH = "${pkgs.rust.packages.stable.rustPlatform.rustLibSrc}";

          # Extra inputs can be added here
          nativeBuildInputs = [
            pkgs.cargo
            pkgs.rustc
            pkgs.rust-analyzer
          ];
        };
      });
```

This works great, I can run the `flake check` and I can also run `nix develop`  and then install the binary to $HOME/.cargo/bin.
I can also run `nix build --no-link path:.` which will make the rustlings command available from `"$(nix path-info .)"/bin/rustlings`.
I would like `nix develop` to automatically do that for me and make the `rustlings` command available in the shell. Is this possible or is crane the wrong direction to take?

Thanks in advance!

**Comment by dpc** [comment, marked as answer]
> I would like nix develop to automatically do that for me and make the rustlings command available in the shell. Is this possible or is crane the wrong direction to take?

`nix develop` is supposed to give you a shell where you can work on the project itself, and not a shell where you project is built.

You can possibly make extra nix shells, and just add `rustlings` into `nativeBuildInputs` there which will make it available. `nix develop .#devShellName` will let you get there.

**Reply by Nequo** [reply]
Great, thank you for the information. A second shell seems like the way to go here! I get that it is unusual to want to do this, but the idea is to be able to clone the repo, run `nix develop .#rustlings` and then run `rustlings watch` and be able to go through the exercises. I think your suggestion will work great.

**Reply by dpc** [reply]
Example: https://github.com/fedimint/fedimint/blob/be13df1517b9369157e60006fccd6318fbae6a31/flake.nix#L479

Here we have a shell just for linting a PR, stripped to save on fetching dependencies in the CI and make it very fast.

So there's no limit in what you can you dev shells for. Flakes are awesome.
